common/defs.go

```diff
@@ -197,58 +199,74 @@ func (p *Policy) Validate() error {
 				}
 			}
 			if len(badRanges) > 0 {
-				errMsg += fmt.Sprintf("The following port ranges are invalid: %s. ", strings.Join(badRanges, ", "))
+				errMsg = append(errMsg, fmt.Sprintf("Rule #%d: The following port ranges are invalid: %s.", ruleNo, strings.Join(badRanges, ", ")))
 			}
 			badPorts := make([]string, 0)
 			for _, port := range r.Ports {
 				if port < 0 || port > MaxPortNumber {
-					badPorts = append(badPorts, string(port))
+					badPorts = append(badPorts, fmt.Sprintf("%d", port))
 				}
 			}
 			if len(badPorts) > 0 {
-				errMsg += fmt.Sprintf("The following ports are invalid: %s. ", strings.Join(badPorts, ", "))
+				errMsg = append(errMsg, fmt.Sprintf("Rule #%d: The following ports are invalid: %s.", ruleNo, strings.Join(badPorts, ", ")))
 			}
 		}
-		if r.Protocol == "icmp" {
+		if r.Protocol != "icmp" {
 			if r.IcmpCode > 0 || r.IcmpType > 0 {
-				errMsg += "ICMP protocol is not specified but ICMP Code and/or ICMP Type are also specified. "
-			} else {
-				if len(r.Ports) > 0 || len(r.PortRanges) > 0 {
-					errMsg += "ICMP protocol is specified but ports are also specified. "
+				errMsg = append(errMsg, fmt.Sprintf("Rule #%d: ICMP protocol is not specified but ICMP Code and/or ICMP Type are also specified.", ruleNo))
+			}
+		} else {
+			if len(r.Ports) > 0 || len(r.PortRanges) > 0 {
+				errMsg = append(errMsg, fmt.Sprintf("Rule #%d: ICMP protocol is specified but ports are also specified.", ruleNo))
+			}
+			if r.IcmpType < 0 || r.IcmpType > MaxIcmpType {
+				errMsg = append(errMsg, fmt.Sprintf("Rule #%d: Invalid ICMP type: %d.", ruleNo, r.IcmpType))
+			}
+			switch r.IcmpType {
+			case 3: // Destination unreachable
```

Comment:
Are those literals (3, 4, 5, 11) not defined by name somewhere in some include file? So that you don't have to specify those values directly?

If not, then I would suggest we define those ICMP types somewhere as constants and then use those.